### ISSUE TITLE
Cannot run multiple backward() in parallel, cpu execution is serialized

### ISSUE THREAD

**Issue by slaweks17** [pytorch/pytorch#17566]
## 🐛 Bug
Hi,
I am running an OMP loop in a libtorch program (Windows 10). 
```
#pragma omp parallel for
      for (int inet=0; inet<NUM_OF_NETS; inet++) { 
        auto& nnStack=*nnStack_arr[inet];
        auto& trainer=*trainers_arr[inet];
        ...
       totalLoss_t.backward();
       ...
}
```
So, separate nets, trainers  prepared earlier. All other objects are locally created within the loop.
Things go as expected, i.e. multithreaded, until I hit 
totalLoss_t.backward();
where things appear serialized (on 12 core machine 8% CPU). totalLoss_t is also a local object.
Once the call is done, the threads appear to start working in parallel again. 
Do I miss some configuration call?
Tried both debug and release builds, the same.

 - PyTorch: 1.0, compiled from latest
 - OS: Windows 10
 - CUDA/cuDNN version: CPU build

**Comment by peterjc123** [comment]
Could you please step into this line and give the stack trace?

**Comment by slaweks17** [comment]
Hi,
Thank you. Here you are:
```
	c10.dll!c10::TensorImpl::data() Line 595	C++
 	c10.dll!c10::TensorImpl::slow_data() Line 611	C++
 	caffe2.dll!at::native::data_ptr(const at::Tensor & self) Line 11	C++
 	caffe2.dll!at::TypeDefault::data_ptr(const at::Tensor & self) Line 4497	C++
 	caffe2.dll!at::Tensor::data_ptr() Line 798	C++
 	caffe2.dll!at::Tensor::data<float>() Line 1347	C++
 	caffe2.dll!at::strided_tensor_iter_fixed<float,8>::strided_tensor_iter_fixed<float,8>(at::Tensor & tensor, bool sort_strides) Line 157	C++
 	caffe2.dll!at::CPU_tensor_apply2<float,float,void <lambda>(float &, const float &) >(at::Tensor tensor1, at::Tensor tensor2, const `anonymous-namespace'::_copy__cpu::__l2::void <lambda>(float &, const float &) op) Line 426	C++
 	caffe2.dll!`anonymous namespace'::_copy__cpu<float,float>(at::Tensor & self, const at::Tensor & src) Line 18	C++
 	caffe2.dll!`anonymous-namespace'::_copy__cpu::__l2::void <lambda>(void)::__l8::<lambda>() Line 23	C++
 	caffe2.dll!`anonymous-namespace'::_copy__cpu::__l2::<lambda>() Line 23	C++
 	caffe2.dll!`anonymous namespace'::_copy__cpu<float>(at::Tensor & self, const at::Tensor & src) Line 26	C++
 	caffe2.dll!at::native::_s_copy__cpu::__l2::void <lambda>(void)::__l8::<lambda>() Line 45	C++
 	caffe2.dll!at::native::_s_copy__cpu::__l2::<lambda>() Line 45	C++
 	caffe2.dll!at::native::_s_copy__cpu(at::Tensor & self, const at::Tensor & src, bool non_blocking) Line 47	C++
 	caffe2.dll!at::CPUFloatType::s_copy_(at::Tensor & self, const at::Tensor & src, bool non_blocking) Line 5056	C++
 	torch.dll!torch::autograd::VariableType::s_copy_(at::Tensor & self, const at::Tensor & src, bool non_blocking) Line 278	C++
 	caffe2.dll!at::TypeDefault::copy_(at::Tensor & self, const at::Tensor & src, bool non_blocking) Line 27	C++
 	caffe2.dll!at::Tensor::copy_(const at::Tensor & src, bool non_blocking) Line 32	C++
 	torch.dll!torch::autograd::generated::`anonymous namespace'::select_backward(at::Tensor grad, c10::ArrayRef<__int64> input_sizes, __int64 dim, __int64 index) Line 625	C++
 	torch.dll!torch::autograd::generated::SelectBackward::apply(std::vector<torch::autograd::Variable,std::allocator<torch::autograd::Variable> > && grads) Line 4538	C++
 	torch.dll!torch::autograd::Function::operator()(std::vector<torch::autograd::Variable,std::allocator<torch::autograd::Variable> > && inputs) Line 116	C++
 	torch.dll!torch::autograd::call_function(torch::autograd::FunctionTask & task) Line 394	C++
 	torch.dll!torch::autograd::Engine::evaluate_function(torch::autograd::FunctionTask & task) Line 425	C++
 	torch.dll!torch::autograd::Engine::thread_main(torch::autograd::GraphTask * graph_task) Line 251	C++
 	torch.dll!torch::autograd::Engine::thread_init(int device) Line 225	C++
```
And another, random break:
```
	c10.dll!c10::StorageImpl::data() Line 134	C++
 	c10.dll!c10::Storage::data() Line 95	C++
 	c10.dll!c10::TensorImpl::storage_initialized() Line 1238	C++
 	caffe2.dll!c10::TensorImpl::data<double>() Line 566	C++
 	caffe2.dll!THDoubleTensor_fill$omp$1() Line 10	C++
 	[External Code]	
 	caffe2.dll!THDoubleTensor_fill(c10::TensorImpl * r_, double value) Line 11	C++
 	caffe2.dll!at::CPUDoubleType::_th_fill_(at::Tensor & self, c10::Scalar value) Line 106	C++
 	caffe2.dll!at::_th_fill_(at::Tensor & self, c10::Scalar value) Line 1550	C++
 	caffe2.dll!at::native::fill_(at::Tensor & self, c10::Scalar value) Line 88	C++
 	caffe2.dll!at::TypeDefault::fill_(at::Tensor & self, c10::Scalar value) Line 2889	C++
 	caffe2.dll!at::Tensor::fill_(c10::Scalar value) Line 279	C++
 	caffe2.dll!at::native::scalar_tensor(c10::Scalar s, const c10::TensorOptions & options) Line 266	C++
 	caffe2.dll!at::TypeDefault::scalar_tensor(c10::Scalar s, const c10::TensorOptions & options) Line 3512	C++
 	caffe2.dll!at::scalar_tensor(c10::Scalar s, const c10::TensorOptions & options) Line 4156	C++
 	caffe2.dll!c10::scalar_to_tensor(c10::Scalar s) Line 13	C++
 	caffe2.dll!at::native::wrapped_scalar_tensor(c10::Scalar scalar) Line 142	C++
 	caffe2.dll!at::native::mul(const at::Tensor & self, c10::Scalar other) Line 164	C++
 	caffe2.dll!at::TypeDefault::mul(const at::Tensor & self, c10::Scalar other) Line 3396	C++
 	torch.dll!torch::autograd::VariableType::mul::__l2::<lambda>() Line 12611	C++
 	torch.dll!torch::autograd::VariableType::mul(const at::Tensor & self, c10::Scalar other) Line 12612	C++
 	caffe2.dll!at::Tensor::mul(c10::Scalar other) Line 434	C++
 	torch.dll!at::operator*(const at::Tensor & x, c10::Scalar y) Line 96	C++
 	torch.dll!torch::autograd::generated::MulBackward1::apply(std::vector<torch::autograd::Variable,std::allocator<torch::autograd::Variable> > && grads) Line 3883	C++
 	torch.dll!torch::autograd::Function::operator()(std::vector<torch::autograd::Variable,std::allocator<torch::autograd::Variable> > && inputs) Line 116	C++
 	torch.dll!torch::autograd::call_function(torch::autograd::FunctionTask & task) Line 394	C++
 	torch.dll!torch::autograd::Engine::evaluate_function(torch::autograd::FunctionTask & task) Line 425	C++
 	torch.dll!torch::autograd::Engine::thread_main(torch::autograd::GraphTask * graph_task) Line 251	C++
 	torch.dll!torch::autograd::Engine::thread_init(int device) Line 225	C++
```

**Comment by slaweku** [comment]
I have just tried it on Mac, same result - backward() is serialized.

**Comment by slaweku** [comment]
Do you think it worth trying to run my program-level multithreading with Intel TBB instead of OpenMP?

**Comment by peterjc123** [comment]
cc @yf225

**Comment by ezyang** [comment]
Oh, I know why this is happening: when you call `backward()`, the work gets shipped out to another thread pool which isn't OMP managed, and that thread pool serializes all CPU execution.

I'm not sure what a good fix for this problem is.

**Comment by albanD** [comment]
Closing this in favor of https://github.com/pytorch/pytorch/issues/18333 that request the same feature.
